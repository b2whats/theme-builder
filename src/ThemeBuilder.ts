import React from 'react'
import { css } from '@emotion/css'
import type { StyleProperties, StateSelectors } from './StyleProperties'
import { objectStyleToString } from './StyleProperties'
import { Tokens } from './tokens'

const Test = () => React.createElement('div')

type MaybeArray<Item> = Item extends []
  ? Item | Item[number]
  : Item[] | Item

export type Slot<Props = unknown> = unknown extends Props
  ? StyleProperties & StateSelectors & { withProps: boolean }
  : Props & { withProps: boolean, component: any }

type ShortHands = 'p' | 'px' | 'py' | 'm' | 'mx' | 'my'
export type ComponentTheme<Props, Slots> = {
  defaultProps?: Partial<Props>
  mapProps?: (props: Props) => Omit<Partial<Props>, ShortHands> | void | false
  slots: {
    [Name in keyof Slots]: {
      type?: 'tag' | 'component',
      styles: Parts<Props, Slots[Name]>[]
    }
  }
}

type ComputeTheme<Theme> = Theme extends ComponentTheme<infer _, infer Slots>
    ? {
      [Name in keyof Slots]: Slots[Name] extends Slot<infer Props>
        ? {
          type: 'component',
          props: Props
        } : {
          type: 'tag',
          className: string,
        } 
    }
    : never
type PropertyAsFunction<Map, Args> = { 
  [Key in keyof Map]?: NonNullable<Map[Key]> extends object
    ? PropertyAsFunction<Map[Key], Args>
    : Map[Key] | ((props: Args) => Map[Key])
}

type Parts<ExternalProps, SlotProps> = SlotProps extends StateSelectors
  ? string
    | PropertyAsFunction<SlotProps, ExternalProps>
    | ((props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string)
  : PropertyAsFunction<SlotProps, ExternalProps>

type SwitchMap<Props, Prop extends keyof Props, Result> = {
  [Value in Extract<Props[Prop], string>]?: Result | ((props: Props) => Result | undefined)
}

interface SlotUtils {
  if<ExternalProps, Props extends MaybeArray<keyof ExternalProps>, Values, PropertyValues extends Values>(
    prop: Props | Props[],
    value: PropertyValues,
    defaultValue?: PropertyValues,
  ): (props: ExternalProps) => Values | undefined
  map<ExternalProps, Prop extends keyof ExternalProps, Values, Default extends Values>(
    prop: Prop,
    body: SwitchMap<ExternalProps, Prop, Values>,
    defaultValue?: Default,
  ): (props: ExternalProps) => Values | undefined
  css<ExternalProps, SlotProps>(
    literals: TemplateStringsArray,
    ...placeholders: Array<(props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string>
  ): (props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string
}

export const utils: SlotUtils = {
  if: (keys, value, defaultValue) => {
    return (props: any) => {
      const isValue = Array.isArray(keys)
        ? keys.every(key => Boolean(props[key]))
        : Boolean(props[keys])
      
      return isValue ? value : defaultValue
    }
  },
  map: (prop, options, defaultValue) => {
    return (props: any) => {
      const value = props[prop]

      return value in options
        ? typeof options[value] === 'function' ? options[value](props) : options[value]
        : defaultValue
    }
  },
  css: (literals, ...placeholders) => {
    return function cssRewrite(props, tokens, slotStyles) {
      let result = ''

      for (let i = 0; i < placeholders.length; i++) {
          result += literals[i];
          result += placeholders[i](props, tokens, slotStyles)
      }
  
      result += literals[literals.length - 1];

      return result
    }
  }
}

const shortHands = {
  m: (value: any) => ({ mt: value, mr: value, mb: value, ml: value }),
  mx: (value: any) => ({ mr: value, ml: value }),
  my: (value: any) => ({ mt: value, mb: value }),
  p: (value: any) => ({ pt: value, pr: value, pb: value, pl: value }),
  px: (value: any) => ({ pr: value, pl: value }),
  py: (value: any) => ({ pt: value, pb: value }),
}

const isObject = (value: any): value is object => value && value.constructor && value.constructor === Object

function expandShorthands<T>(
  styles: T
): T {
  let results = {} as T

  if (!isObject(styles)) return styles

  for (let [key, value] of Object.entries(styles)) {
    if (shortHands[key]) {
      Object.assign(results, shortHands[key](value))

      continue
    }
    if (['disabled', 'hover', 'active', 'visited', 'checked'].includes(key)) {
      results[key] = expandShorthands(value)
    }

    results[key] = value
  }

  return results
}

const weight = (val: any): number => {
  if (typeof val === 'string') return 1
  if (typeof val === 'object') return 2
  if (typeof val === 'function') return 3

  return 0
}

export function mergeStyles<Styles extends Parts<any, any>>(arr: Styles[]): Styles[] {
  return arr.sort((a, b) => weight(a) - weight(b)).reduce<Styles[]>((acc, cur) => {
    const prev = acc[acc.length - 1]
    if (typeof prev !== typeof cur || typeof cur === 'function') {
      acc.push(cur)

      return acc
    }

    if (typeof prev === 'string' && typeof cur === 'string') {
      (acc[acc.length - 1] as string) = prev + cur

      return acc
    }
    if (typeof prev === 'object' && typeof cur === 'object') {
      (acc[acc.length - 1] as object) = {
        ...prev as object,
        ...cur as object
      }

      return acc
    }

    return acc
  }, [])

}

export class ThemeBuilder<Theme extends ComponentTheme<any, any>> {
  theme = {
    slots: {}
  } as Theme

  slot<Name extends keyof this['theme']['slots']>(
    name: Name,
    styles: this['theme']['slots'][Name]['styles'] | this['theme']['slots'][Name]['styles'][number],
    type?: 'component' | 'tag'
  ) {
    this.theme.slots[name] = { 
      type,
      styles: Array.isArray(styles)
        ? mergeStyles((styles as []).map(expandShorthands))
        : [expandShorthands(styles)]
    }

    return this
  }
  defaultProps(props: this['theme']['defaultProps']) {
    if (!props) return this

    this.theme.defaultProps = expandShorthands(props)

    return this
  }
  mapProps(map: this['theme']['mapProps']) {
    this.theme.mapProps = map

    return this
  }
  merge(builder: this | undefined) {
    if (!builder) return this

    const instance = new ThemeBuilder<Theme>()

    if (this.theme.defaultProps || builder.theme.defaultProps) {
      instance.defaultProps(
        Object.assign({}, this.theme.defaultProps, builder.theme.defaultProps)
      )
    }

    if (this.theme.mapProps || builder.theme.mapProps) {
      instance.mapProps((props) => ({
        ...this.theme.mapProps && this.theme.mapProps(props),
        ...builder.theme.mapProps && builder.theme.mapProps(props),
      }))
    }

    for (let [name, parts] of Object.entries(this.theme.slots)) {
      instance.slot(name, mergeStyles([
        ...parts.styles,
        ...builder.theme.slots[name] ? builder.theme.slots[name].styles : [],
      ]), parts.type)
    }

    return instance
  }
  compute(props: this['theme']['defaultProps'], tokens: Tokens): ComputeTheme<this['theme']> {
    const result = {} as any

    props = {
      ...this.theme.defaultProps,
      ...expandShorthands(props),
    }

    if (this.theme.mapProps) {
      Object.assign(props, this.theme.mapProps(props))
    }
    
    for (let [name, { type, styles }] of Object.entries(this.theme.slots)) {
      if (type === 'tag') {
        let slotStyles = {}

        const className = styles.map(style => {
          if (typeof style === 'string') {
            return css(style)
          }
  
          if (typeof style === 'object') {
            if (style.withProps) {
              style = {
                ...style,
                ...props,
              }
            }
            slotStyles = style
            return css(objectStyleToString(props, tokens, style as Slot))
          }
  
          if (typeof style === 'function') {
            return css(style(props, tokens, slotStyles))
          }
  
          return ''
        }).join(' ')

        result[name] = {
          type,
          className,
        }
      } else {
        const componentProps = styles.map(style => {
          if (typeof style === 'object') {
            if (style.withProps) {
              style = {
                ...style,
                ...props,
              }

              delete style.withProps
            }

            const result = {}
            for (let prop in style) {
              result[prop] = typeof style[prop] === 'function'
                ? style[prop](props)
                : style[prop]
            }

            return result
          }

          return {}
        }).reduce((acc, curr) => {
          Object.assign(acc, curr)

          return acc
        })

        result[name] = {
          type,
          props: componentProps as any,
        }
      }

    }

    return result
  }
}




interface ButtonProps extends StyleProperties {
  /** Иконка слева */
  children?: any
  /** Иконка слева */
  iconBefore?: React.ReactNode
  /** Иконка справа */
  iconAfter?: React.ReactNode
  /** Размер кнопки */
  size?: 's' | 'm' | 'l'
  /** Имя кнопки */
  name?: string
  /** Значение кнопки */
  value?: string | number
  /** Внешний вид кнопки */
  kind?: 'fill' | 'outline' | 'flat'
  /** Нажатое состояние кнопки */
  checked?: boolean
  /** Действие которое совершает кнопка */
  type?: 'button' | 'submit' | 'reset'
  /** Кнопка со спиннером */
  loading?: boolean
  /** Квадратная круглая и вытянутая форма с закругленными углами */
  shape?: 'pill' | 'square' | 'circle'
  /** Вторая строка */
  multiline?: boolean
  /** Кнопка занимает всю ширину */
  block?: boolean
  /** Ссылка для перехода */
  href?: string

  pressedOffset?: number
}


interface TextProps extends StyleProperties {
  /** Размер текста фолбек */
  fontSize?: number
  /** Коэффициент для высоты строки */
  lineHeight?: number | 'none' | 'inherit'
  /** Уплотненная высота строки */
  dense?: boolean
  /** Расстояние между буквами */
  letterSpacing?: number
  /** Жирное начертание */
  bold?: boolean
  /** Легкое начертание */
  light?: boolean
  /** Курсивное начертание */
  italic?: boolean
  /** Текст в верхнем регистре */
  uppercase?: boolean
  /** Подчеркнутый текст */
  underline?: 'dashed' | 'dotted' | boolean
  /** Обрезка высоты строки у текста сверху и снизу */
  crop?: boolean
  /** Текст без переносов */
  noWrap?: boolean
  /** Текст с сохранением всех пробелов */
  pre?: boolean
  /** Многоточие в конце строки */
  truncate?: boolean
  /** Инлайновое поведение */
  inline?: boolean
  /** Блочное поведение */
  block?: boolean
  /** Пресет компонента */
  preset?: 'title' | 'title-small' | 'heading-large' | 'heading' | 'heading-small'
  preset2?: 'title' | 'title-small' | 'heading-large' | 'heading' | 'heading-small'
}

type ButtonTheme = ComponentTheme<ButtonProps, {
  Button: Slot,
  Text: Slot<TextProps>,
}>


const a = new ThemeBuilder<ButtonTheme>()
  .defaultProps({
    size: 'm',
    type: 'button',
  })
  .mapProps(({ shape }) => { if (false) return { pt: 1 } })
  .slot('Button', {
    shrink: false
  })
  .slot('Button', {
    withProps: true,
    shrink: false,
    grow: (props) => true,
    color: (props) => 'black20',
  })
  .slot('Button', [
    '',
    {
      pl: utils.if('iconAfter', 12),
      pr: utils.if(['iconAfter', 'children'], 12, 16),
      color: utils.map('kind', {
        fill: utils.map('size', { s: 'red400', m: 'green400' }),
        flat: 'black20'
      }, 'red300')
    },
    utils.css`
      color: ${(props, tokens) => ''};
    `
  ])
  .slot('Text', {
    withProps: true,
    bold: () => true,
    component: Test
  })

  const b = new ThemeBuilder<ButtonTheme>()
  .defaultProps({
    size: 'm',
    type: 'button',
  })
  .mapProps(({ shape }) => { if (false) return {} })
  .slot('Button', [
    `    
      font-family: inherit;
      cursor: pointer;
      text-align: center;
      text-decoration: none;
      margin: 0;
      line-height: 0;
      position: relative;
      white-space: nowrap;
      -webkit-tap-highlight-color: rgba(0,0,0,0);
      -webkit-touch-callout: none;
      &::-moz-focus-inner {
        border: 0;
      }
      a& {
        display: inline-flex;
        justify-content: center;
        align-items: center;
      }
    `,
    {
      borderRadius: 5,
      pl: ({ iconBefore, children }) => iconBefore && children ? 12 : 16,
      pr: ({ iconBefore, children }) => iconBefore && children ? 12 : 16,
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: 'transparent',
      focus: false,
      minHeight: ({ size }) => size,
      py: ({ size }) => ({ s: 4, m: 6, l: 8 }[size!]),
      userSelect: false,
    },
    (props, tokens, slotStyle) => ''
  ])

