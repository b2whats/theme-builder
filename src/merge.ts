import { isObject, isFunction, isString } from './utils'

const emptyCache = {}
const cache: Map<object, Map<object, object>> = new Map()

export const mergeTheme = (a: object = emptyCache, b: object = emptyCache) => {
  let cacheA = cache.get(a)

  if (cacheA === undefined) {
    cacheA = new Map()
    cache.set(a, cacheA)
  }

  let cacheB = cacheA.get(b)

  if (cacheB === undefined) {
    cacheB = mergeObject(a, b, {
      mapProps: (a, b) => {
        if (isFunction(a) && isFunction(b)) {
          return (props: any) => ({
            ...a(props),
            ...b(props),
          })
        }
        return b
      },
      asString: (a, b) => {
        if (isString(a) && isString(b)) {
          return a + b
        }
        return b
      },
    }, { removeEmpty: true })
    cacheA.set(b, cacheB)
  }

  return cacheB
}

function cloneUnlessOtherwiseSpecified(value: any) {
	return isObject(value)
		? mergeObject({}, value)
		: value
}

type CustomMerge = {
  [key: string]: (a: unknown, b: unknown) => any,
}
type OptionsMerge = {
  removeEmpty?: any,
}

function mergeObject(target: object, source: object, customMerge: CustomMerge = {}, options: OptionsMerge = {}): Record<string, any> {
  const result = {}

  Object.keys(target).forEach((key) => {
    result[key] = cloneUnlessOtherwiseSpecified(target[key])
  })

  Object.keys(source).forEach((key) => {
    const targetValue = target[key]
    const sourceValue = source[key]
    
    if (customMerge[key]) {
      result[key] = customMerge[key](targetValue, sourceValue)
    } else if (isObject(targetValue) && isObject(sourceValue)) {
      result[key] = mergeObject(targetValue, sourceValue, customMerge, options)
    } else {
      result[key] = cloneUnlessOtherwiseSpecified(sourceValue)
    }
    
    if (options.removeEmpty && result[key] == null) {
      delete result[key]
    }
  })

  return result
}