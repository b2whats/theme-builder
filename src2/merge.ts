import { isObject, isThemeBuilder } from './utils'

function cloneUnlessOtherwiseSpecified(value: any) {
	return isObject(value)
		? mergeObject({}, value)
		: value
}

export function mergeObject(target: object, source: object): Record<string, any> {
  const result = {}

  Object.keys(target).forEach((key) => {
    result[key] = cloneUnlessOtherwiseSpecified(target[key])
  })

  Object.keys(source).forEach((key) => {
    const targetValue = target[key]
    const sourceValue = source[key]
    
    if (isThemeBuilder(targetValue) && isThemeBuilder(sourceValue)) {
      result[key] = targetValue.merge(sourceValue)
    } else if (isObject(targetValue) && isObject(sourceValue)) {
      result[key] = mergeObject(targetValue, sourceValue)
    } else {
      result[key] = cloneUnlessOtherwiseSpecified(sourceValue)
    }
  })

  return result
}

type UnaryFn<Arg, Return> = (arg: Arg) => Return

export const weakMemoize = <Arg extends object, Return>(func: UnaryFn<Arg, Return>): UnaryFn<Arg, Return> => {
  let cache = new WeakMap()
  
  return arg => {
    if (cache.has(arg)) return cache.get(arg)

    let result = func(arg)
    cache.set(arg, result)
    
    return result
  }
}

export const primitiveMemoize = <Arg extends string | number | null | undefined, Return>(func: UnaryFn<Arg, Return>): UnaryFn<Arg, Return> => {
  let cache = {} as any
  
  return arg => {
    if (cache[arg]) return cache[arg]

    let result = func(arg)
    cache[arg] = result
    
    return result
  }
}

export const memoize = <Arg extends any, Return>(func: UnaryFn<Arg, Return>): UnaryFn<Arg, Return> => {
  let cache = new Map()
  
  return arg => {
    if (arg === undefined) return func(arg)

    const hash = JSON.stringify(arg)

    if (cache.has(hash)) return cache.get(hash)

    let result = func(arg)
    cache.set(hash, result)
    
    return result
  }
}

const createCacheWithTheme = weakMemoize(outerTheme => {
  return weakMemoize(theme => {
    if (isObject(outerTheme) && isObject(theme)) {
      return mergeObject(outerTheme, theme)
    } else {
      return outerTheme
    }
  })
})

export function mergeTheme<T extends object>(target: T, source?: object): T {
  // return createCacheWithTheme(target)(source) as T
  return '' as any
}