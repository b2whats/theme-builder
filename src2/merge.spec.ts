import { mergeObject, mergeTheme } from './merge'
import { ThemeBuilder, ComponentTheme, Slot } from './ThemeBuilder'

type ButtonTheme = ComponentTheme<{}, {
  Button: Slot,
}>

type TextTheme = ComponentTheme<{}, {
  Text: Slot,
}>

const ButtonTheme = new ThemeBuilder<ButtonTheme>()
  .slot('Button', {
    width: 100
  })

const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
  .slot('Button', {
    width: 200,
    height: 100,
  })
const TextTheme = new ThemeBuilder<TextTheme>()
  .slot('Text', {
    width: 200
  })

const target: object = {
  string: 'string',
  array: [1,2],
  number: 1,
  nestedObject: {
    object: {
      l: 1,
      m: 2,
      s: 3,
    },
    string: 'string',
    array: [1,2],
  },
  object: {
    l: 1,
    m: 2,
    s: 3,
  },
}

const source: object = {
  string: 'string 2',
  array: [11],
  number: 11,
  nestedObject: {
    object: {
      l: 11,
    },
    string: 'string 2',
    array: [22],
    additionalNumber: 11
  },
  object: {
    s: 33,
  },
  additionalString: 'additionalString'
}

const targetWithComponent: object = {
  ...target,
  Button: ButtonTheme
}

const sourceWithComponent: object = {
  ...source,
  Text: TextTheme
}


test('mergeObject', () => {
  const merge = mergeObject(target, source)

  expect(merge).toMatchSnapshot()
})

test('mergeObject - simple', () => {
  let firstArg = { one: 1, merge: 1 }
  let secondArg = { two: 2, merge: 2 }

  let firstResult = mergeTheme(firstArg, secondArg)
  let secondResult = mergeTheme(firstArg, secondArg)

  expect(firstResult).toBe(secondResult)
  expect(firstResult).toMatchSnapshot()

  let newArg = { two: 2, merge: 2 }
  let newResult = mergeTheme(firstArg, newArg)

  expect(newResult).not.toBe(firstResult)
  expect(newResult).toMatchSnapshot()
})

test('mergeTheme - with component', () => {
  let firstArg = { Button: ButtonTheme }
  let secondArg = { Text: TextTheme }

  let firstResult = mergeTheme(firstArg, secondArg)
  expect(firstResult).toMatchSnapshot()
})

test('mergeTheme - tokens with component', () => {
  let firstArg = { Button: ButtonTheme, one: 1 }
  let secondArg = { Text: TextTheme, one: '1', two: '2' }

  let firstResult = mergeTheme(firstArg, secondArg)
  expect(firstResult).toMatchSnapshot()
})

test('mergeTheme - tokens with same component', () => {
  let firstArg = { Button: ButtonTheme, one: 1 }
  let secondArg = { Button: ButtonTheme2, one: '1', two: '2' }

  let firstResult = mergeTheme(firstArg, secondArg)
  expect(firstResult).toMatchSnapshot()
})