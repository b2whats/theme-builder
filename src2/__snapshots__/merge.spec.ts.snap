// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeObject - simple 1`] = `
Object {
  "merge": 2,
  "one": 1,
  "two": 2,
}
`;

exports[`mergeObject - simple 2`] = `
Object {
  "merge": 2,
  "one": 1,
  "two": 2,
}
`;

exports[`mergeObject 1`] = `
Object {
  "additionalString": "additionalString",
  "array": Array [
    11,
  ],
  "nestedObject": Object {
    "additionalNumber": 11,
    "array": Array [
      22,
    ],
    "object": Object {
      "l": 11,
      "m": 2,
      "s": 3,
    },
    "string": "string 2",
  },
  "number": 11,
  "object": Object {
    "l": 1,
    "m": 2,
    "s": 33,
  },
  "string": "string 2",
}
`;

exports[`mergeTheme - tokens with component 1`] = `
Object {
  "Button": ThemeBuilder {
    "compute": [Function],
    "merge": [Function],
    "theme": Object {
      "slots": Object {
        "Button": Object {
          "styles": Array [
            Object {
              "width": 100,
            },
          ],
          "type": undefined,
        },
      },
    },
  },
  "Text": ThemeBuilder {
    "compute": [Function],
    "merge": [Function],
    "theme": Object {
      "slots": Object {
        "Text": Object {
          "styles": Array [
            Object {
              "width": 200,
            },
          ],
          "type": undefined,
        },
      },
    },
  },
  "one": "1",
  "two": "2",
}
`;

exports[`mergeTheme - tokens with same component 1`] = `
Object {
  "Button": ThemeBuilder {
    "compute": [Function],
    "merge": [Function],
    "theme": Object {
      "slots": Object {
        "Button": Object {
          "styles": Array [
            Object {
              "height": 100,
              "width": 200,
            },
          ],
          "type": undefined,
        },
      },
    },
  },
  "one": "1",
  "two": "2",
}
`;

exports[`mergeTheme - with component 1`] = `
Object {
  "Button": ThemeBuilder {
    "compute": [Function],
    "merge": [Function],
    "theme": Object {
      "slots": Object {
        "Button": Object {
          "styles": Array [
            Object {
              "width": 100,
            },
          ],
          "type": undefined,
        },
      },
    },
  },
  "Text": ThemeBuilder {
    "compute": [Function],
    "merge": [Function],
    "theme": Object {
      "slots": Object {
        "Text": Object {
          "styles": Array [
            Object {
              "width": 200,
            },
          ],
          "type": undefined,
        },
      },
    },
  },
}
`;
